relax type equality constraint

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61031 remove UsageError in favor of ValueError
* #60639 change language from array to tensor
* **#60638 relax type equality constraint**

Initial proposal in https://github.com/pytorch/pytorch/pull/58981#issuecomment-866690334. Opposed to the proposal, this PR only allows relaxing the type equality constraint to a common superclass constraint, for example `torch.Tensor` vs `torch.nn.Parameter`. Inputs that do not share a common superclass will still fail.

Differential Revision: [D29626811](https://our.internmc.facebook.com/intern/diff/D29626811)

pmeier committed:
relax type equality constraint

Initial proposal in https://github.com/pytorch/pytorch/pull/58981#issuecomment-866690334. Opposed to the proposal, this PR only allows relaxing the type equality constraint to a common superclass constraint, for example `torch.Tensor` vs `torch.nn.Parameter`. Inputs that do not share a common superclass will still fail.

[ghstack-poisoned]

pmeier committed:
Update on "relax type equality constraint"


Initial proposal in https://github.com/pytorch/pytorch/pull/58981#issuecomment-866690334. Opposed to the proposal, this PR only allows relaxing the type equality constraint to a common superclass constraint, for example `torch.Tensor` vs `torch.nn.Parameter`. Inputs that do not share a common superclass will still fail.

[ghstack-poisoned]

pmeier committed:
Update on "relax type equality constraint"


Initial proposal in https://github.com/pytorch/pytorch/pull/58981#issuecomment-866690334. Opposed to the proposal, this PR only allows relaxing the type equality constraint to a common superclass constraint, for example `torch.Tensor` vs `torch.nn.Parameter`. Inputs that do not share a common superclass will still fail.

[ghstack-poisoned]

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -577,31 +577,43 @@ def _to_tensor(array_or_scalar_like: Any) -> Tuple[Optional[_TestingErrorMeta],
+        return None
+
+    msg_fmtstr = f"Except for Python scalars, {{}}, but got {type(actual)} and {type(expected)} instead."
+    common_superclass = isinstance(actual, type(expected)) or isinstance(expected, type(actual))
```
It's not that the object share a superclass, but that one object has the other's class as a superclass.

mruberry replied:
Also: let's add a comment here explaining how isinstance works with sub/super classes.

pmeier replied:
> Also: let's add a comment here explaining how isinstance works with sub/super classes.

Not sure I get the point? If one knows what `isinstance` does, the implementation is straight forward. If not one should read the documentation?

pmeier replied:
> It's not that the object share a superclass, but that one object has the other's class as a superclass.

I went with the "directly related" terminology, because "one object has the other's class as a superclass." is also not completely right: both objects can have the same time (arguably the more common case) and this check would still pass.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -577,31 +577,43 @@ def _to_tensor(array_or_scalar_like: Any) -> Tuple[Optional[_TestingErrorMeta],
+    msg_fmtstr = f"Except for Python scalars, {{}}, but got {type(actual)} and {type(expected)} instead."
+    common_superclass = isinstance(actual, type(expected)) or isinstance(expected, type(actual))
+    if not common_superclass:
+        return _TestingErrorMeta(AssertionError, msg_fmtstr.format("inputs are required to share a common superclass"))
```
This error message should reflect the value of `allow_subclasses`, too

If `allow_subclasses=False` the two objects must have the same type

If `allows_subclasses=True` then one object must be a subclass of the other's type

pmeier replied:
I've only added the `allow_subclasses=False` in the error message, since "one object must be a subclass of the other's type" is true regardless of `allow_subclasses`.

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -577,31 +577,43 @@ def _to_tensor(array_or_scalar_like: Any) -> Tuple[Optional[_TestingErrorMeta],
     Args:
         actual (Any): Actual array-or-scalar-like.
         expected (Any): Expected array-or-scalar-like.
+        allow_subclasses (bool): If ``True`` inputs of a common superclass are allowed. Otherwise type equality is
```
I think this doc needs to be revised because, again, any two Python objects have a common superclass (object).

mruberry replied:
Wait, shouldn't this say:

"If this is \`\`False\`\` (the default) then both inputs must have the same type. If \`\`True\`\` then one input may be a (direct or indirect) subclass of the other."

?

pmeier committed:
Update on "relax type equality constraint"


Initial proposal in https://github.com/pytorch/pytorch/pull/58981#issuecomment-866690334. Opposed to the proposal, this PR only allows relaxing the type equality constraint to a common superclass constraint, for example `torch.Tensor` vs `torch.nn.Parameter`. Inputs that do not share a common superclass will still fail.

[ghstack-poisoned]

pmeier committed:
Update on "relax type equality constraint"


Initial proposal in https://github.com/pytorch/pytorch/pull/58981#issuecomment-866690334. Opposed to the proposal, this PR only allows relaxing the type equality constraint to a common superclass constraint, for example `torch.Tensor` vs `torch.nn.Parameter`. Inputs that do not share a common superclass will still fail.

[ghstack-poisoned]

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -734,15 +749,17 @@ def assert_close(
-    they are considered close if their structure matches and all their elements are considered close according to the
-    above definition.
+    :attr:`actual` and :attr:`expected` can be :class:`~torch.Tensor`'s or any tensor-or-scalar-like that share the a
+    common superclass (if :attr:`allow_subclasses` is ``True``), from which :class:`torch.Tensor`'s can be constructed
```
I think this would be "any tensor-like or scalar-like as long as :attr:\`actual\` and :attr:\`expected\` are the same types or one is a (direct or indirect) subclass of the other."?

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -766,19 +783,21 @@ def assert_close(
         UsageError: If only :attr:`rtol` or :attr:`atol` is specified.
-        AssertionError: If corresponding array-likes have different types.
+        AssertionError: If corresponding tensor-or-scalar-likes do not share a common superclass.
+        AssertionError: If :attr:`allow_subclasses` is ``False``, but tensor-or-scalar-likes have different types.
```
"but actual and expected are not Python numbers and have different types"?

mruberry commented on torch/testing/_asserts.py:
```diff
@@ -766,19 +783,21 @@ def assert_close(
         UsageError: If a :class:`torch.Tensor` can't be constructed from an array-or-scalar-like.
         UsageError: If only :attr:`rtol` or :attr:`atol` is specified.
-        AssertionError: If corresponding array-likes have different types.
+        AssertionError: If corresponding tensor-or-scalar-likes do not share a common superclass.
```
"If neither actual or expected are Python numbers, share the same type, or one is not a (direct or indirect) subclass of the other"?

mruberry commented:
Language on this one is tricky but i think it still needs a tweak. I'm going to import it and continue reviewing the rest of the stack so it's ready to merge, however.

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626811).

pmeier committed:
Update on "relax type equality constraint"


Initial proposal in https://github.com/pytorch/pytorch/pull/58981#issuecomment-866690334. Opposed to the proposal, this PR only allows relaxing the type equality constraint to a common superclass constraint, for example `torch.Tensor` vs `torch.nn.Parameter`. Inputs that do not share a common superclass will still fail.

Differential Revision: [D29626811](https://our.internmc.facebook.com/intern/diff/D29626811)

[ghstack-poisoned]

pmeier commented:
@mruberry 

> i think it still needs a tweak

Thanks for catching this again. I simply missed these places in my first round of addressing your comments. 

I went with the "directly related" terminology, since IMO it is just as expressive as "same types or one is a (direct or indirect) subclass of the other." but a lot more concise.

mruberry reviewed (approved):
"Directly related" seems fine for now

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626811).
